lib/samlp.js

```diff
@@ -180,6 +238,105 @@ function generateInstant() {
   return date.getUTCFullYear() + '-' + ('0' + (date.getUTCMonth()+1)).slice(-2) + '-' + ('0' + date.getUTCDate()).slice(-2) + 'T' + ('0' + date.getUTCHours()).slice(-2) + ":" + ('0' + date.getUTCMinutes()).slice(-2) + ":" + ('0' + date.getUTCSeconds()).slice(-2) + "Z";
 }
 
+// Validates an XML that contains a Signature element (signature inside XML)
+function validateXmlSignature(xml, options, callback) {
+  var calculatedThumbprint = '';
+  var signature = options.signature || xpath.select(signaturePath, xml)[0];
+  if (!signature){
+    return callback(new Error('Signature is missing (xpath: ' + signaturePath + ')'));
+  }
+
+  if (options.thumprints){
+    // Make sure thumprints is an array for the validation 
+    options.thumbprints = options.thumbprints instanceof Array ? options.thumbprints : [options.thumbprints];
+  }
+
+  var sig = new SignedXml(null, { idAttribute: 'AssertionID' });
+
+  sig.keyInfoProvider = {
+    getKeyInfo: function () {
+      return "<X509Data></X509Data>";
+    },
+    getKey: function (keyInfo) {
+
+      //If there's no embedded signing cert, use the configured cert through options
+      if(!keyInfo || keyInfo.length===0){
+        if(!options.cert) throw new Error('options.cert must be specified for SAMLResponses with no embedded signing certificate');
+        return certToPEM(options.cert);
+      }
+
+      //If there's an embedded signature and thumprints are provided check that
+      if (options.thumbprints && options.thumbprints.length > 0)  {
```

Comment:
Is it ok to prioritize the embedded cert instead of the configured cert `options.cert` ?